File: 矩阵/54.螺旋矩阵.go
package 矩阵

func spiralOrder(matrix [][]int) []int {
	m, n := len(matrix), len(matrix[0])
	res := []int{}
	l, r, t, d := 0, n-1, 0, m-1
	for true {
		for i := l; i <= r; i++ {
			res = append(res, matrix[t][i])
		}
		if t++; t > d {
			break
		}
		for i := t; i <= d; i++ {
			res = append(res, matrix[i][r])
		}
		if r--; r < l {
			break
		}
		for i := r; i >= l; i-- {
			res = append(res, matrix[d][i])
		}
		if d--; d < t {
			break
		}
		for i := d; i >= t; i-- {
			res = append(res, matrix[i][l])
		}
		if l++; l > r {
			break
		}
	}
	return res
}

File: 矩阵/54.螺旋矩阵.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length,n=matrix[0].length;
        int t=0,l=0,r=n-1,d=m-1;
        List<Integer> list = new ArrayList<>();
        while(true){
            for(int i=l;i<=r;i++){
                list.add(matrix[t][i]);
            }
            if(++t>d) break;
            for(int i=t;i<=d;i++){
                list.add(matrix[i][r]);
            }
            if(--r<l)break;
            for(int i=r;i>=l;i--){
                list.add(matrix[d][i]);
            }
            if(--d<t)break;
            for(int i=d;i>=t;i--){
                list.add(matrix[i][l]);
            }
            if(++l>r)break;
        }
        return list;
    }
}